src/connection.rs

```diff
@@ -658,24 +658,22 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
                 let sender = self.stream_sender.clone();
                 Stream::new(stream_id, self.id, config, credit, credit, sender)
             };
-            let window_update;
+            let mut window_update = None;
             {
                 let mut shared = stream.shared();
                 if is_finish {
                     shared.update_state(self.id, stream_id, State::RecvClosed);
                 }
                 shared.window = shared.window.saturating_sub(frame.body_len());
                 shared.buffer.push(frame.into_body());
-                if !is_finish
-                    && shared.window == 0
-                    && self.config.window_update_mode == WindowUpdateMode::OnReceive
-                {
-                    shared.window = self.config.receive_window;
-                    let mut frame = Frame::window_update(stream_id, self.config.receive_window);
-                    frame.header_mut().ack();
-                    window_update = Some(frame)
-                } else {
-                    window_update = None
+
+                if matches!(self.config.window_update_mode, WindowUpdateMode::OnReceive) {
+                    if let Some(credit) = shared.next_window_update() {
```

Comment:
> I understand that the former is the context and the latter is the currently configured WindowUpdateMode

The argument given is just what you otherwise check against at the call-site, `WindowUpdateMode::OnReceive` in this case, i.e. you just leave the check against the config to `next_window_update()`. It just strikes me as a bit odd that you have to check the configured window update mode before calling `next_window_update()` when that function itself is already checking the configured mode in order to compute the update _and_ may even return `None` if there is no next window update. I thus thought it may be better to have `next_window_update()` decide a) whether there is a next window update for the given mode (returning `None` if not) and b) how to generate it, simultaneously. I guess if `next_window_update()` did not return an `Option`, making it clear that the call-site needs to decide whether to generate one, I would not have made the suggestion.  I don't really mind though, it was just a thought, I don't feel strongly about it either.

Reply:
Thanks for the detailed elaboration. I am torn between the two, finding both not fully intuitive. I will leave it as is, but happy to discuss alternative approaches.